pgjdbc/src/test/java/org/postgresql/test/jdbc2/PropertyLoaderTest.java

```diff
@@ -0,0 +1,117 @@
+    verifyUrl(ldr, "jdbc:postgresql://localhost/test", "localhost", "5432", "test");
+    verifyUrl(ldr, "jdbc:postgresql://localhost:5432/test", "localhost", "5432", "test");
+    verifyUrl(ldr, "jdbc:postgresql://127.0.0.1/anydbname", "127.0.0.1", "5432", "anydbname");
+    verifyUrl(ldr, "jdbc:postgresql://127.0.0.1:5433/hidden", "127.0.0.1", "5433", "hidden");
+    verifyUrl(ldr, "jdbc:postgresql://[::1]:5740/db", "[::1]", "5740", "db");
+
+    // failover urls
+    verifyUrl(ldr, "jdbc:postgresql://localhost,127.0.0.1:5432/test", "localhost,127.0.0.1",
+        "5432,5432", "test");
+    verifyUrl(ldr, "jdbc:postgresql://localhost:5433,127.0.0.1:5432/test", "localhost,127.0.0.1",
+        "5433,5432", "test");
+    verifyUrl(ldr, "jdbc:postgresql://[::1],[::1]:5432/db", "[::1],[::1]", "5432,5432", "db");
+    verifyUrl(ldr, "jdbc:postgresql://[::1]:5740,127.0.0.1:5432/db", "[::1],127.0.0.1", "5740,5432",
+        "db");
+  }
+
+  private void verifyUrl(PropertyLoader ldr, String url, String hosts, String ports, String dbName) throws Exception {
+    Method parseMethod = ldr.getClass().getDeclaredMethod("parseURL", String.class, Properties.class);
+    parseMethod.setAccessible(true);
+    Properties p = (Properties) parseMethod.invoke(ldr, url, null);
+    assertTrue(url != null);
+    assertEquals(url, dbName, p.getProperty(PGProperty.PG_DBNAME.getName()));
+    assertEquals(url, hosts, p.getProperty(PGProperty.PG_HOST.getName()));
+    assertEquals(url, ports, p.getProperty(PGProperty.PG_PORT.getName()));
+  }
+
+  @Test
+  public void testEnvironment() throws Exception {
+    Map<String, String> env = new HashMap();
+    env.put("PGHOST", "localhost");
+    env.put("PGPORT", "5432");
+    env.put("PGUSER", "me");
+    env.put("PGAPPNAME", "myapp");
+
+    Properties props = new Properties();
+    PropertyLoader.loadEnv(props, env);
+
+    assertTrue(PGProperty.PG_HOST.isPresent(props));
+    assertEquals("localhost", PGProperty.PG_HOST.get(props));
+
+    assertTrue(PGProperty.PG_PORT.isPresent(props));
+    assertEquals(5432, PGProperty.PG_PORT.getInt(props));
+
+    assertTrue(PGProperty.USER.isPresent(props));
+    assertEquals("me", PGProperty.USER.get(props));
+  }
+
+  @Test
+  public void testOrder() throws Exception {
+    Map<String, String> env = new HashMap();
+    env.put("PGHOST", "fromenv");
+    env.put("PGUSER", "fromenv");  // To override with URL
+    env.put("PGDATABASE", "fromenv");  // To override with info
+
+    Properties info = new Properties();
+    PGProperty.PG_DBNAME.set(info, "frominfo");  // Override env
+    PGProperty.USER.set(info, "frominfo");  // To override with URL
+
+    String url = "jdbc:postgresql:?user=fromurl";
+    Properties props = PropertyLoader.load(url, info, env);
+
+    assertNotNull(props);
+
+    System.out.println(props.get(PGProperty.PG_HOST.getName()));
+    System.out.println(props.getProperty(PGProperty.PG_HOST.getName()));
```

Comment:
@davecramer @grzm these two lines outputs like this:

```
null
fromenv
```

at https://travis-ci.org/pgjdbc/pgjdbc/jobs/395419870#L831

I'm lost :/

Reply:
Ya, and this is the problem with this whole thing, there are multiple places where variables are set. url, properties, now env. Your specific problem is in parseURL it returns a new properties object and on line 46 ``` if ((props = parseURL(url, props)) == null) {```` overrides all of the work done in the previous lines...

Reply:
Does it mean that `Properties props = new Properties(defaults);` loose information ?

I don't understeand why `.get` returns `null` while `.getProperty` find the value.

Reply:
`.get` is implemented in `java.util.Hashtable`, and `Properties` is a derived class.
It would be nice if `Properties` could override `get` and check for defaults, however it is not the case.

That is why `getProperty` checks for defaults and `get` just has no idea of their existence.

Reply:
getProperty will return the default and return it if the value is null. 

https://github.com/pgjdbc/pgjdbc/pull/1222/files/2f1499cfaf56d2bd27913054a0fd2a5eaa3d0ad7#diff-73c1698d73484995a2026fad2a31fcaaR46 is where you are losing all the work above

Reply:
@vlsi ok, i understand !

Reply:
> Ya, and this is the problem with this whole thing, there are multiple places where variables are set.

About this. I understand this must be done right. I'm aware of this, and this is why a added a unit test for testing the order of loading the parameters and prevent regression. Do you trust this ? @grzm I'll be glad for your review on this specific part. I didn't find source order test in pique.

Since it's tricky, it's a good candidate for shared, tested code instead of having each application writing it's own version. And with opt-in, it is safe for the user-base, isn't-it ?

Anyway, I wish you're happy to have `Driver.java` below 600 sloc :-)